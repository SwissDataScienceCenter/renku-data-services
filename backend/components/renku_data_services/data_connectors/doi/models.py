"""Models for DOIs."""

from dataclasses import dataclass
from typing import Any

from pydantic import BaseModel, Field


@dataclass(frozen=True, eq=True, kw_only=True)
class DOIMetadata:
    """Model for DOI metadata."""

    name: str
    description: str
    keywords: list[str]


class InvenioRecordMetadata(BaseModel):
    """Representation of a record's metadata."""

    title: str | None = Field(default=None)
    description: str | None = Field(default=None)
    keywords: list[str] | None = Field(default=None)


class InvenioRecord(BaseModel):
    """Schema for the representation of a record from the InvenioRDM API."""

    metadata: InvenioRecordMetadata | None = Field(default=None)


class DataverseMetadataBlockCitationField(BaseModel):
    """A metadata field of citation metadata."""

    type_name: str = Field(alias="typeName")
    multiple: bool = Field()
    type_class: str = Field(alias="typeClass")
    value: Any = Field()  # TODO: can we find better types here?


class DataverseMetadataBlockCitation(BaseModel):
    """Representation of citation metadata."""

    fields: list[DataverseMetadataBlockCitationField] = Field(default_factory=list)


class DataverseMetadataBlocks(BaseModel):
    """Represents metadata of a Dataverse dataset."""

    citation: DataverseMetadataBlockCitation | None = Field()


class DataverseDatasetVersion(BaseModel):
    """Representation of a dataset version."""

    metadata_blocks: DataverseMetadataBlocks | None = Field(alias="metadataBlocks")


class DataverseDataset(BaseModel):
    """Representation of a dataset in Dataverse."""

    latest_version: DataverseDatasetVersion | None = Field(alias="latestVersion")


class DataverseDatasetResponse(BaseModel):
    """DataverseDatasetResponse is returned by the Dataverse dataset API."""

    status: str = Field()
    data: DataverseDataset | None = Field()
