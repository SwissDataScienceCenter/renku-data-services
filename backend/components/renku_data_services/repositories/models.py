"""Models for connected services."""

from dataclasses import dataclass

from ulid import ULID


@dataclass(frozen=True, eq=True, kw_only=True)
class RepositoryPermissions:
    """Repository permissions for git operations."""

    pull: bool
    push: bool

    @classmethod
    def default(cls) -> "RepositoryPermissions":
        """Default permissions."""
        return cls(pull=False, push=False)


@dataclass(frozen=True, eq=True, kw_only=True)
class RepositoryMetadata:
    """Repository metadata."""

    etag: str | None
    git_http_url: str
    web_url: str
    permissions: RepositoryPermissions


@dataclass(frozen=True, eq=True, kw_only=True)
class RepositoryProviderMatch:
    """Repository provider match data."""

    provider_id: str
    connection_id: ULID | None
    repository_metadata: RepositoryMetadata | None
