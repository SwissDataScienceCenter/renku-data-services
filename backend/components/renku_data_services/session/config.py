"""Configuration for session module."""

import os
from dataclasses import dataclass
from datetime import timedelta

from pydantic import ValidationError as PydanticValidationError

from renku_data_services.app_config import logging
from renku_data_services.session import crs as session_crs

logger = logging.getLogger(__name__)


@dataclass
class BuildsConfig:
    """Configuration for container image builds."""

    enabled: bool = False
    build_output_image_prefix: str | None = None
    vscodium_python_run_image: str | None = None
    build_strategy_name: str | None = None
    push_secret_name: str | None = None
    buildrun_retention_after_failed: timedelta | None = None
    buildrun_retention_after_succeeded: timedelta | None = None
    buildrun_build_timeout: timedelta | None = None
    node_selector: dict[str, str] | None = None
    tolerations: list[session_crs.Toleration] | None = None

    @classmethod
    def from_env(cls) -> "BuildsConfig":
        """Create a config from environment variables."""
        enabled = os.environ.get("IMAGE_BUILDERS_ENABLED", "false").lower() == "true"
        build_output_image_prefix = os.environ.get("BUILD_OUTPUT_IMAGE_PREFIX")
        vscodium_python_run_image = os.environ.get("BUILD_VSCODIUM_PYTHON_RUN_IMAGE")
        build_strategy_name = os.environ.get("BUILD_STRATEGY_NAME")
        push_secret_name = os.environ.get("BUILD_PUSH_SECRET_NAME")
        buildrun_retention_after_failed_seconds = int(os.environ.get("BUILD_RUN_RETENTION_AFTER_FAILED_SECONDS") or "0")
        buildrun_retention_after_failed = (
            timedelta(seconds=buildrun_retention_after_failed_seconds)
            if buildrun_retention_after_failed_seconds > 0
            else None
        )
        buildrun_retention_after_succeeded_seconds = int(
            os.environ.get("BUILD_RUN_RETENTION_AFTER_SUCCEEDED_SECONDS") or "0"
        )
        buildrun_retention_after_succeeded = (
            timedelta(seconds=buildrun_retention_after_succeeded_seconds)
            if buildrun_retention_after_succeeded_seconds > 0
            else None
        )
        buildrun_build_timeout_seconds = int(os.environ.get("BUILD_RUN_BUILD_TIMEOUT") or "0")
        buildrun_build_timeout = (
            timedelta(seconds=buildrun_build_timeout_seconds) if buildrun_build_timeout_seconds > 0 else None
        )

        if os.environ.get("DUMMY_STORES", "false").lower() == "true":
            enabled = True  # Enable image builds when running tests

        node_selector: dict[str, str] | None = None
        node_selector_str = os.environ.get("BUILD_NODE_SELECTOR")
        if node_selector_str:
            try:
                node_selector = session_crs.NodeSelector.model_validate_json(node_selector_str).root
            except PydanticValidationError:
                logger.error("Could not validate BUILD_NODE_SELECTOR. Will not use node selector for image builds.")

        tolerations: list[session_crs.Toleration] | None = None
        tolerations_str = os.environ.get("BUILD_NODE_TOLERATIONS")
        if tolerations_str:
            try:
                tolerations = session_crs.Tolerations.model_validate_json(tolerations_str).root
            except PydanticValidationError:
                logger.error("Could not validate BUILD_NODE_TOLERATIONS. Will not use tolerations for image builds.")

        return cls(
            enabled=enabled or False,
            build_output_image_prefix=build_output_image_prefix or None,
            vscodium_python_run_image=vscodium_python_run_image or None,
            build_strategy_name=build_strategy_name or None,
            push_secret_name=push_secret_name or None,
            buildrun_retention_after_failed=buildrun_retention_after_failed,
            buildrun_retention_after_succeeded=buildrun_retention_after_succeeded,
            buildrun_build_timeout=buildrun_build_timeout,
            node_selector=node_selector,
            tolerations=tolerations,
        )
