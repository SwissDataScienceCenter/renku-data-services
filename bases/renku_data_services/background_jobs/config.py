"""Configurations for background jobs."""

import os
from collections.abc import Callable
from dataclasses import dataclass

from sqlalchemy.ext.asyncio import AsyncSession, create_async_engine
from sqlalchemy.orm import sessionmaker

from renku_data_services.authz.authz import Authz
from renku_data_services.authz.config import AuthzConfig
from renku_data_services.data_connectors.migration_utils import DataConnectorMigrationTool
from renku_data_services.errors import errors
from renku_data_services.message_queue.config import RedisConfig
from renku_data_services.message_queue.db import EventRepository
from renku_data_services.message_queue.redis_queue import RedisQueue
from renku_data_services.namespace.db import GroupRepository
from renku_data_services.project.db import ProjectRepository
from renku_data_services.users.db import UsersSync
from renku_data_services.users.kc_api import IKeycloakAPI, KeycloakAPI


@dataclass
class SyncConfig:
    """Main configuration."""

    syncer: UsersSync
    kc_api: IKeycloakAPI
    authz_config: AuthzConfig
    group_repo: GroupRepository
    event_repo: EventRepository
    project_repo: ProjectRepository

    # NEW
    data_connector_migration_tool: DataConnectorMigrationTool

    session_maker: Callable[..., AsyncSession]

    @classmethod
    def from_env(cls, prefix: str = "") -> "SyncConfig":
        """Generate a configuration from environment variables."""
        pg_host = os.environ.get(f"{prefix}DB_HOST", "localhost")
        pg_user = os.environ.get(f"{prefix}DB_USER", "renku")
        pg_port = os.environ.get(f"{prefix}DB_PORT", "5432")
        db_name = os.environ.get(f"{prefix}DB_NAME", "renku")
        pg_password = os.environ.get(f"{prefix}DB_PASSWORD")
        if pg_password is None:
            raise errors.ConfigurationError(
                message="Please provide a database password in the 'DB_PASSWORD' environment variable."
            )
        async_sqlalchemy_url = f"postgresql+asyncpg://{pg_user}:{pg_password}@{pg_host}:{pg_port}/{db_name}"
        # NOTE: the pool here is not used to serve HTTP requests, it is only used in background jobs.
        # Therefore, we want to consume very few connections and we can wait for an available connection
        # much longer than the default 30 seconds. In our tests syncing 15 users times out with the default.
        engine = create_async_engine(async_sqlalchemy_url, pool_size=4, max_overflow=0, pool_timeout=600)
        session_maker: Callable[..., AsyncSession] = sessionmaker(engine, class_=AsyncSession, expire_on_commit=False)  # type: ignore[call-overload]
        redis = RedisConfig.from_env(prefix)
        message_queue = RedisQueue(redis)

        authz_config = AuthzConfig.from_env()
        event_repo = EventRepository(session_maker=session_maker, message_queue=message_queue)
        group_repo = GroupRepository(
            session_maker,
            event_repo=event_repo,
            group_authz=Authz(authz_config),
            message_queue=message_queue,
        )
        project_repo = ProjectRepository(
            session_maker=session_maker,
            message_queue=message_queue,
            event_repo=event_repo,
            group_repo=group_repo,
            authz=Authz(authz_config),
        )

        # NEW
        data_connector_migration_tool = DataConnectorMigrationTool(
            session_maker=session_maker,
            authz=Authz(authz_config),
        )

        syncer = UsersSync(
            session_maker,
            message_queue=message_queue,
            event_repo=event_repo,
            group_repo=group_repo,
            authz=Authz(authz_config),
        )
        keycloak_url = os.environ[f"{prefix}KEYCLOAK_URL"]
        client_id = os.environ[f"{prefix}KEYCLOAK_CLIENT_ID"]
        client_secret = os.environ[f"{prefix}KEYCLOAK_CLIENT_SECRET"]
        realm = os.environ.get(f"{prefix}KEYCLOAK_REALM", "Renku")
        kc_api = KeycloakAPI(keycloak_url=keycloak_url, client_id=client_id, client_secret=client_secret, realm=realm)
        return cls(
            syncer,
            kc_api,
            authz_config,
            group_repo,
            event_repo,
            project_repo,
            data_connector_migration_tool,
            session_maker,
        )
