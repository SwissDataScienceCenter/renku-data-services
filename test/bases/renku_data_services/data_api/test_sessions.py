"""Tests for sessions blueprints."""

from asyncio import AbstractEventLoop
from collections.abc import AsyncIterator, Coroutine
from typing import Any

import pytest
import pytest_asyncio
from pytest import FixtureRequest
from sanic_testing.testing import SanicASGITestClient, TestingResponse

from renku_data_services.app_config.config import Config
from renku_data_services.crc.apispec import ResourcePool
from renku_data_services.users.models import UserInfo


@pytest.fixture
def create_session_environment(sanic_client: SanicASGITestClient, admin_headers):
    async def create_session_environment_helper(name: str, **payload) -> dict[str, Any]:
        payload = payload.copy()
        payload.update({"name": name})
        payload["description"] = payload.get("description") or "A session environment."
        payload["container_image"] = payload.get("container_image") or "some_image:some_tag"

        _, res = await sanic_client.post("/api/data/environments", headers=admin_headers, json=payload)

        assert res.status_code == 201, res.text
        assert res.json is not None
        return res.json

    return create_session_environment_helper


@pytest.fixture
def create_session_launcher(sanic_client: SanicASGITestClient, user_headers):
    async def create_session_launcher_helper(name: str, project_id: str, **payload) -> dict[str, Any]:
        payload = payload.copy()
        payload.update({"name": name, "project_id": project_id})
        payload["description"] = payload.get("description") or "A session launcher."
        if "environment" not in payload:
            payload["environment"] = {
                "environment_kind": "CUSTOM",
                "name": "Test",
                "container_image": "some_image:some_tag",
            }

        _, res = await sanic_client.post("/api/data/session_launchers", headers=user_headers, json=payload)

        assert res.status_code == 201, res.text
        assert res.json is not None
        return res.json

    return create_session_launcher_helper


@pytest.fixture
def launch_session(
    sanic_client: SanicASGITestClient,
    user_headers: dict,
    regular_user: UserInfo,
    app_config: Config,
    request: FixtureRequest,
    event_loop: AbstractEventLoop,
):
    async def launch_session_helper(
        payload: dict, headers: dict = user_headers, user: UserInfo = regular_user
    ) -> TestingResponse:
        _, res = await sanic_client.post("/api/data/sessions", headers=headers, json=payload)
        assert res.status_code == 201, res.text
        assert res.json is not None
        assert "name" in res.json
        session_id: str = res.json.get("name", "unknown")

        def cleanup():
            event_loop.run_until_complete(app_config.nb_config.k8s_v2_client.delete_server(session_id, user.id))

        # request.addfinalizer(cleanup)
        return res

    return launch_session_helper


@pytest.mark.asyncio
async def test_get_all_session_environments(
    sanic_client: SanicASGITestClient, unauthorized_headers, create_session_environment
) -> None:
    await create_session_environment("Environment 1")
    await create_session_environment("Environment 2")
    await create_session_environment("Environment 3")

    _, res = await sanic_client.get("/api/data/environments", headers=unauthorized_headers)

    assert res.status_code == 200, res.text
    assert res.json is not None
    environments = res.json
    assert {env["name"] for env in environments} == {
        "Environment 1",
        "Environment 2",
        "Environment 3",
    }


@pytest.mark.asyncio
async def test_get_session_environment(
    sanic_client: SanicASGITestClient, unauthorized_headers, create_session_environment
) -> None:
    env = await create_session_environment(
        "Environment 1",
        description="Some environment.",
        container_image="test_image:latest",
    )
    environment_id = env["id"]

    _, res = await sanic_client.get(f"/api/data/environments/{environment_id}", headers=unauthorized_headers)

    assert res.status_code == 200, res.text
    assert res.json is not None
    assert res.json.get("name") == "Environment 1"
    assert res.json.get("description") == "Some environment."
    assert res.json.get("container_image") == "test_image:latest"


@pytest.mark.asyncio
async def test_post_session_environment(sanic_client: SanicASGITestClient, admin_headers) -> None:
    payload = {
        "name": "Environment 1",
        "description": "A session environment.",
        "container_image": "some_image:some_tag",
    }

    _, res = await sanic_client.post("/api/data/environments", headers=admin_headers, json=payload)

    assert res.status_code == 201, res.text
    assert res.json is not None
    assert res.json.get("name") == "Environment 1"
    assert res.json.get("description") == "A session environment."
    assert res.json.get("container_image") == "some_image:some_tag"


@pytest.mark.asyncio
async def test_post_session_environment_unauthorized(sanic_client: SanicASGITestClient, user_headers) -> None:
    payload = {
        "name": "Environment 1",
        "description": "A session environment.",
        "container_image": "some_image:some_tag",
    }

    _, res = await sanic_client.post("/api/data/environments", headers=user_headers, json=payload)

    assert res.status_code == 401, res.text


@pytest.mark.asyncio
async def test_patch_session_environment(
    sanic_client: SanicASGITestClient, admin_headers, create_session_environment
) -> None:
    env = await create_session_environment("Environment 1")
    environment_id = env["id"]

    payload = {
        "name": "New name",
        "description": "New description.",
        "container_image": "new_image:new_tag",
    }

    _, res = await sanic_client.patch(f"/api/data/environments/{environment_id}", headers=admin_headers, json=payload)

    assert res.status_code == 200, res.text
    assert res.json is not None
    assert res.json.get("name") == "New name"
    assert res.json.get("description") == "New description."
    assert res.json.get("container_image") == "new_image:new_tag"


@pytest.mark.asyncio
async def test_patch_session_environment_unauthorized(
    sanic_client: SanicASGITestClient, user_headers, create_session_environment
) -> None:
    env = await create_session_environment("Environment 1")
    environment_id = env["id"]

    payload = {
        "name": "New name",
        "description": "New description.",
        "container_image": "new_image:new_tag",
    }

    _, res = await sanic_client.patch(f"/api/data/environments/{environment_id}", headers=user_headers, json=payload)

    assert res.status_code == 403, res.text


@pytest.mark.asyncio
async def test_delete_session_environment(
    sanic_client: SanicASGITestClient, admin_headers, create_session_environment
) -> None:
    env = await create_session_environment("Environment 1")
    environment_id = env["id"]

    _, res = await sanic_client.delete(f"/api/data/environments/{environment_id}", headers=admin_headers)

    assert res.status_code == 204, res.text


@pytest.mark.asyncio
async def test_delete_session_environment_unauthorized(
    sanic_client: SanicASGITestClient, user_headers, create_session_environment
) -> None:
    env = await create_session_environment("Environment 1")
    environment_id = env["id"]

    _, res = await sanic_client.delete(f"/api/data/environments/{environment_id}", headers=user_headers)

    assert res.status_code == 403, res.text


@pytest.mark.asyncio
async def test_get_all_session_launchers(
    sanic_client: SanicASGITestClient,
    user_headers,
    create_project,
    create_session_launcher,
) -> None:
    project_1 = await create_project("Project 1")
    project_2 = await create_project("Project 2")

    await create_session_launcher("Launcher 1", project_id=project_1["id"])
    await create_session_launcher("Launcher 2", project_id=project_2["id"])
    await create_session_launcher("Launcher 3", project_id=project_2["id"])

    _, res = await sanic_client.get("/api/data/session_launchers", headers=user_headers)

    assert res.status_code == 200, res.text
    assert res.json is not None
    launchers = res.json
    assert {launcher["name"] for launcher in launchers} == {
        "Launcher 1",
        "Launcher 2",
        "Launcher 3",
    }


@pytest.mark.asyncio
async def test_get_session_launcher(
    sanic_client: SanicASGITestClient,
    unauthorized_headers,
    create_project,
    create_session_environment,
    create_session_launcher,
) -> None:
    project = await create_project("Some project", visibility="public")
    env = await create_session_environment("Some environment")
    launcher = await create_session_launcher(
        "Some launcher",
        project_id=project["id"],
        description="Some launcher.",
        environment={"id": env["id"]},
    )
    launcher_id = launcher["id"]

    _, res = await sanic_client.get(f"/api/data/session_launchers/{launcher_id}", headers=unauthorized_headers)

    assert res.status_code == 200, res.text
    assert res.json is not None
    assert res.json.get("name") == "Some launcher"
    assert res.json.get("project_id") == project["id"]
    assert res.json.get("description") == "Some launcher."
    environment = res.json.get("environment", {})
    assert environment.get("environment_kind") == "GLOBAL"
    assert environment.get("id") == env["id"]
    assert environment.get("container_image") == env["container_image"]
    assert res.json.get("resource_class_id") is None


@pytest.mark.asyncio
async def test_get_project_launchers(
    sanic_client: SanicASGITestClient,
    user_headers,
    create_project,
    create_session_launcher,
) -> None:
    project_1 = await create_project("Project 1")
    project_2 = await create_project("Project 2")

    await create_session_launcher("Launcher 1", project_id=project_1["id"])
    await create_session_launcher("Launcher 2", project_id=project_2["id"])
    await create_session_launcher("Launcher 3", project_id=project_2["id"])

    _, res = await sanic_client.get(f"/api/data/projects/{project_2['id']}/session_launchers", headers=user_headers)

    assert res.status_code == 200, res.text
    assert res.json is not None
    launchers = res.json
    assert {launcher["name"] for launcher in launchers} == {"Launcher 2", "Launcher 3"}


@pytest.mark.asyncio
async def test_post_session_launcher(
    sanic_client: SanicASGITestClient,
    valid_resource_pool_payload: dict[str, Any],
    user_headers,
    admin_headers,
    member_1_headers,
    create_project,
    create_resource_pool,
) -> None:
    project = await create_project("Some project")
    resource_pool_data = valid_resource_pool_payload

    resource_pool = await create_resource_pool(admin=True, **resource_pool_data)

    payload = {
        "name": "Launcher 1",
        "project_id": project["id"],
        "description": "A session launcher.",
        "resource_class_id": resource_pool["classes"][0]["id"],
        "environment": {
            "container_image": "some_image:some_tag",
            "name": "custom_name",
            "environment_kind": "CUSTOM",
        },
    }

    _, res = await sanic_client.post("/api/data/session_launchers", headers=admin_headers, json=payload)

    assert res.status_code == 201, res.text
    assert res.json is not None
    assert res.json.get("name") == "Launcher 1"
    assert res.json.get("project_id") == project["id"]
    assert res.json.get("description") == "A session launcher."
    environment = res.json.get("environment", {})
    assert environment.get("environment_kind") == "CUSTOM"
    assert environment.get("container_image") == "some_image:some_tag"
    assert environment.get("id") is not None
    assert res.json.get("resource_class_id") == resource_pool["classes"][0]["id"]


@pytest.mark.asyncio
async def test_post_session_launcher_unauthorized(
    sanic_client: SanicASGITestClient,
    valid_resource_pool_payload: dict[str, Any],
    user_headers,
    admin_headers,
    create_project,
    create_resource_pool,
    regular_user,
    create_session_environment,
) -> None:
    project = await create_project("Some project")
    resource_pool_data = valid_resource_pool_payload
    resource_pool_data["public"] = False

    resource_pool = await create_resource_pool(admin=True, **resource_pool_data)
    environment = await create_session_environment("Test environment")

    payload = {
        "name": "Launcher 1",
        "project_id": project["id"],
        "description": "A session launcher.",
        "resource_class_id": resource_pool["classes"][0]["id"],
        "environment": {"id": environment["id"]},
    }

    _, res = await sanic_client.post("/api/data/session_launchers", headers=user_headers, json=payload)

    assert res.status_code == 403, res.text


@pytest.mark.asyncio
async def test_delete_session_launcher(
    sanic_client: SanicASGITestClient,
    user_headers,
    create_project,
    create_session_launcher,
) -> None:
    project = await create_project("Some project")
    launcher = await create_session_launcher("Launcher 1", project_id=project["id"])
    launcher_id = launcher["id"]

    _, res = await sanic_client.delete(f"/api/data/session_launchers/{launcher_id}", headers=user_headers)

    assert res.status_code == 204, res.text


@pytest.mark.asyncio
async def test_patch_session_launcher(
    sanic_client: SanicASGITestClient,
    valid_resource_pool_payload: dict[str, Any],
    user_headers,
    create_project,
    create_resource_pool,
) -> None:
    project = await create_project("Some project 1")
    resource_pool_data = valid_resource_pool_payload
    resource_pool = await create_resource_pool(admin=True, **resource_pool_data)

    payload = {
        "name": "Launcher 1",
        "project_id": project["id"],
        "description": "A session launcher.",
        "resource_class_id": resource_pool["classes"][0]["id"],
        "environment": {
            "container_image": "some_image:some_tag",
            "name": "custom_name",
            "environment_kind": "CUSTOM",
        },
    }

    _, res = await sanic_client.post("/api/data/session_launchers", headers=user_headers, json=payload)

    assert res.status_code == 201, res.text
    assert res.json is not None
    assert res.json.get("name") == "Launcher 1"
    assert res.json.get("description") == "A session launcher."
    environment = res.json.get("environment", {})
    assert environment.get("environment_kind") == "CUSTOM"
    assert environment.get("container_image") == "some_image:some_tag"
    assert environment.get("id") is not None
    assert res.json.get("resource_class_id") == resource_pool["classes"][0]["id"]

    patch_payload = {
        "name": "New Name",
        "description": "An updated session launcher.",
        "resource_class_id": resource_pool["classes"][1]["id"],
    }
    _, res = await sanic_client.patch(
        f"/api/data/session_launchers/{res.json['id']}", headers=user_headers, json=patch_payload
    )
    assert res.status_code == 200, res.text
    assert res.json is not None
    assert res.json.get("name") == patch_payload["name"]
    assert res.json.get("description") == patch_payload["description"]
    assert res.json.get("resource_class_id") == patch_payload["resource_class_id"]


@pytest.mark.asyncio
async def test_patch_session_launcher_environment(
    sanic_client: SanicASGITestClient,
    valid_resource_pool_payload: dict[str, Any],
    user_headers,
    create_project,
    create_resource_pool,
    create_session_environment,
) -> None:
    project = await create_project("Some project 1")
    resource_pool_data = valid_resource_pool_payload
    resource_pool = await create_resource_pool(admin=True, **resource_pool_data)
    global_env = await create_session_environment("Some environment")

    # Create a new custom environment with the launcher
    payload = {
        "name": "Launcher 1",
        "project_id": project["id"],
        "description": "A session launcher.",
        "resource_class_id": resource_pool["classes"][0]["id"],
        "environment": {
            "container_image": "some_image:some_tag",
            "name": "custom_name",
            "environment_kind": "CUSTOM",
        },
    }
    _, res = await sanic_client.post("/api/data/session_launchers", headers=user_headers, json=payload)
    assert res.status_code == 201, res.text
    assert res.json is not None
    environment = res.json.get("environment", {})
    assert environment.get("environment_kind") == "CUSTOM"
    assert environment.get("container_image") == "some_image:some_tag"
    assert environment.get("id") is not None

    # Patch in a global environment
    patch_payload = {
        "environment": {"id": global_env["id"]},
    }
    _, res = await sanic_client.patch(
        f"/api/data/session_launchers/{res.json['id']}", headers=user_headers, json=patch_payload
    )
    assert res.status_code == 200, res.text
    assert res.json is not None
    launcher_id = res.json["id"]
    global_env["environment_kind"] = "GLOBAL"
    assert res.json["environment"] == global_env

    # Trying to patch a field of the global environment should fail
    patch_payload = {
        "environment": {"container_image": "new_image"},
    }
    _, res = await sanic_client.patch(
        f"/api/data/session_launchers/{launcher_id}", headers=user_headers, json=patch_payload
    )
    assert res.status_code == 422, res.text

    # Patching in a wholly new custom environment over the global is allowed
    patch_payload = {
        "environment": {"container_image": "new_image", "name": "new_custom", "environment_kind": "CUSTOM"},
    }
    _, res = await sanic_client.patch(
        f"/api/data/session_launchers/{launcher_id}", headers=user_headers, json=patch_payload
    )
    assert res.status_code == 200, res.text


@pytest.fixture
def anonymous_user_headers() -> dict[str, str]:
    return {"Renku-Auth-Anon-Id": "some-random-value-1234"}


@pytest.mark.asyncio
async def test_starting_session_anonymous(
    sanic_client: SanicASGITestClient,
    create_project,
    create_session_launcher,
    user_headers,
    app_config: Config,
    admin_headers,
    launch_session,
    anonymous_user_headers,
) -> None:
    _, res = await sanic_client.post(
        "/api/data/resource_pools",
        json=ResourcePool.model_validate(app_config.default_resource_pool, from_attributes=True).model_dump(
            mode="json", exclude_none=True
        ),
        headers=admin_headers,
    )
    assert res.status_code == 201, res.text
    project: dict[str, Any] = await create_project(
        "Some project",
        visibility="public",
        repositories=["https://github.com/SwissDataScienceCenter/renku-data-services"],
    )
    launcher: dict[str, Any] = await create_session_launcher(
        "Launcher 1",
        project_id=project["id"],
        environment={
            "container_image": "renku/renkulab-py:3.10-0.23.0-amalthea-sessions-3",
            "environment_kind": "CUSTOM",
            "name": "test",
            "port": 8888,
        },
    )
    launcher_id = launcher["id"]
    project_id = project["id"]
    payload = {"project_id": project_id, "launcher_id": launcher_id}
    session_res = await launch_session(payload, headers=anonymous_user_headers)
    _, res = await sanic_client.get(f"/api/data/sessions/{session_res.json['name']}", headers=anonymous_user_headers)
    assert res.status_code == 200, res.text
    assert res.json["name"] == session_res.json["name"]
    _, res = await sanic_client.get("/api/data/sessions", headers=anonymous_user_headers)
    assert res.status_code == 200, res.text
    assert len(res.json) > 0
    assert session_res.json["name"] in [i["name"] for i in res.json]
    # Should be able to patch some fields of the custom environment
    patch_payload = {
        "environment": {"container_image": "nginx:latest"},
    }
    _, res = await sanic_client.patch(
        f"/api/data/session_launchers/{launcher_id}", headers=user_headers, json=patch_payload
    )
    assert res.status_code == 200, res.text
    assert res.json["environment"]["container_image"] == "nginx:latest"
