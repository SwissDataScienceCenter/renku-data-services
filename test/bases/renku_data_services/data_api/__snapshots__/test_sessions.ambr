# serializer version: 1
# name: test_get_all_session_environments
  list([
    dict({
      'args': list([
        '/entrypoint.sh jupyter server --ServerApp.ip=0.0.0.0 --ServerApp.port=8888 --ServerApp.base_url=$RENKU_BASE_URL_PATH --ServerApp.token="" --ServerApp.password="" --ServerApp.allow_remote_access=true --ContentsManager.allow_hidden=true --ServerApp.allow_origin=* --ServerApp.root_dir="/home/jovyan/work"',
      ]),
      'command': list([
        'sh',
        '-c',
      ]),
      'container_image': 'renku/renkulab-py:latest',
      'default_url': '/lab',
      'description': 'Standard python environment',
      'gid': 100,
      'is_archived': False,
      'mount_directory': '/home/jovyan/work',
      'name': 'Python/Jupyter',
      'port': 8888,
      'strip_path_prefix': False,
      'uid': 1000,
      'working_directory': '/home/jovyan/work',
    }),
    dict({
      'args': list([
        '/entrypoint.sh jupyter server --ServerApp.ip=0.0.0.0 --ServerApp.port=8888 --ServerApp.base_url=$RENKU_BASE_URL_PATH --ServerApp.token="" --ServerApp.password="" --ServerApp.allow_remote_access=true --ContentsManager.allow_hidden=true --ServerApp.allow_origin=* --ServerApp.root_dir="/home/jovyan/work"',
      ]),
      'command': list([
        'sh',
        '-c',
      ]),
      'container_image': 'renku/renkulab-r:latest',
      'default_url': '/rstudio',
      'description': 'Standard R environment',
      'gid': 100,
      'is_archived': False,
      'mount_directory': '/home/jovyan/work',
      'name': 'Rstudio',
      'port': 8888,
      'strip_path_prefix': False,
      'uid': 1000,
      'working_directory': '/home/jovyan/work',
    }),
    dict({
      'container_image': 'some_image:some_tag',
      'default_url': '/lab',
      'description': 'A session environment.',
      'gid': 1000,
      'is_archived': False,
      'name': 'Environment 1',
      'port': 8080,
      'strip_path_prefix': False,
      'uid': 1000,
    }),
    dict({
      'container_image': 'some_image:some_tag',
      'default_url': '/lab',
      'description': 'A session environment.',
      'gid': 1000,
      'is_archived': False,
      'name': 'Environment 2',
      'port': 8080,
      'strip_path_prefix': False,
      'uid': 1000,
    }),
    dict({
      'container_image': 'some_image:some_tag',
      'default_url': '/lab',
      'description': 'A session environment.',
      'gid': 1000,
      'is_archived': False,
      'name': 'Environment 3',
      'port': 8080,
      'strip_path_prefix': False,
      'uid': 1000,
    }),
    dict({
      'container_image': 'some_image:some_tag',
      'default_url': '/lab',
      'description': 'A session environment.',
      'gid': 1000,
      'is_archived': True,
      'name': 'Environment 4',
      'port': 8080,
      'strip_path_prefix': False,
      'uid': 1000,
    }),
  ])
# ---
# name: test_get_all_session_launchers
  list([
    dict({
      'description': 'A session launcher.',
      'environment': dict({
        'container_image': 'some_image:some_tag',
        'default_url': '/lab',
        'environment_image_source': 'image',
        'environment_kind': 'CUSTOM',
        'gid': 1000,
        'is_archived': False,
        'name': 'Test',
        'port': 8080,
        'strip_path_prefix': False,
        'uid': 1000,
      }),
      'name': 'Launcher 1',
    }),
    dict({
      'description': 'A session launcher.',
      'environment': dict({
        'container_image': 'some_image:some_tag',
        'default_url': '/lab',
        'environment_image_source': 'image',
        'environment_kind': 'CUSTOM',
        'gid': 1000,
        'is_archived': False,
        'name': 'Test',
        'port': 8080,
        'strip_path_prefix': False,
        'uid': 1000,
      }),
      'name': 'Launcher 2',
    }),
    dict({
      'description': 'A session launcher.',
      'environment': dict({
        'container_image': 'some_image:some_tag',
        'default_url': '/lab',
        'environment_image_source': 'image',
        'environment_kind': 'CUSTOM',
        'gid': 1000,
        'is_archived': False,
        'name': 'Test',
        'port': 8080,
        'strip_path_prefix': False,
        'uid': 1000,
      }),
      'name': 'Launcher 3',
    }),
  ])
# ---
